src/shogun/machine/GeneralizedLinearMachine.cpp

```diff
@@ -0,0 +1,299 @@
+{
+	SGVector<float64_t> z = compute_z(X, w, bias);
+	SGVector<float64_t> result = non_linearity(z);
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::compute_z(SGMatrix<float64_t> X, SGVector<float64_t> w, float64_t bias)
+{
+	SGVector<float64_t> z = linalg::matrix_prod(X, w, false);
+	return z;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::non_linearity(SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		float64_t bias = 0;
+
+		if(m_fit_intercept)
+			bias = (1 - m_eta) * exp(m_eta);
+
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				result[i] = z[i] * exp(m_eta) + bias;
+			else
+				result[i] = exp(z[i]);
+		}
+		break;
+	
+	default:
+		break;
+	}
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::gradient_non_linearity(SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				
+				result[i] = exp(m_eta);
+			else
+				result[i] = exp(z[i]);
+		}
+		break;
+	
+	default:
+		break;
+	}
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::compute_grad_L2_loss_w(SGMatrix<float64_t> X, SGVector<float64_t> y, SGVector<float64_t> w, float64_t bias)
+{
+	int32_t n_samples = y.vlen;
+	int32_t n_features = X.num_rows;
+	if(m_tau == NULL)
+		m_tau = SGMatrix<float64_t>::create_identity_matrix(w.vlen, 1.0);
+	SGMatrix<float64_t> inv_cov = linalg::matrix_prod(m_tau, m_tau, true, false);
+	
+	SGVector<float64_t> z = compute_z(X, w, bias);
+	SGVector<float64_t> mu = non_linearity(z);
+	SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+	SGVector<float64_t> grad_w(w.vlen);
+	switch (distribution)
+	{
+	case POISSON:
+		//PYTHON CODE
+		//grad_w = ((np.dot(grad_mu.T, X) -
+        //              np.dot((y * grad_mu / mu).T, X)).T)
+		break;
+	
+	default:
+		break;
+	}
+	for (int i = 0; i < grad_w.vlen; i++)
+		grad_w[i] *= 1. / n_samples;
+	
+	linalg::add(grad_w, linalg::matrix_prod(inv_cov, w, false), 1.0, m_lambda * (1 - m_alpha));
+
+	return grad_w;
+}
+
+float64_t GeneralizedLinearMachine::compute_grad_L2_loss_bias(SGMatrix<float64_t> X, SGVector<float64_t> y, SGVector<float64_t> w, float64_t bias)
+{
+	int32_t n_samples = y.vlen;
+	SGVector<float64_t> z = compute_z(X, w, bias);
+	SGVector<float64_t> mu = non_linearity(z);
+	SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+	float64_t grad_bias = 0;
+	switch (distribution)
+	{
+	case POISSON:
+		for (int i = 0; i < grad_mu.vlen; i++)
+		{
+			grad_bias += grad_mu[i];
+			grad_bias -= y[i] * grad_mu[i] / mu[i];
+		}
+		break;
+	
+	default:
+		break;
+	}
+	grad_bias *= 1. / n_samples;
+
+	return grad_bias;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::apply_proximal_operator(SGVector<float64_t> w, float64_t threshold)
```

Comment:
where is this code coming from?

Reply:
This code is written in C++ by me entirely but the calculations are based on a python repository [PyGLMNet](https://github.com/glm-tools/pyglmnet/blob/master/pyglmnet/pyglmnet.py) which has The MIT License (MIT). Is that a problem?

Reply:
no calculations are fine, no need to reference those (all the math is pretty standard). It just came as a surprise to me that this huge blob of code appears out of nowhere. Especially this proximal operator. Could you explain what it does, how it fits into the rest of the code?

Reply:
Here the proximal operator applies a soft thresholding step for k≠0 after every update iteration:(There is a tiny error in the if statement, I'll fix it)
Basically, if the weight is smaller than a particular threshold, we equate it's value to zero. This helps to reduce the number of variables according to the elastic-net penalty.

The elastic net interpolates between two extremes. When α=0 the penalized model is known as ridge regression and when α=1 it is known as LASSO.

Reply:
ok! Good